internal/cache/theme.go

```diff
@@ -0,0 +1,108 @@
+package cache
+
+import (
+	"strconv"
+
+	"fyne.io/fyne/v2"
+	"fyne.io/fyne/v2/internal/svg"
+)
+
+var overrides = make(map[fyne.Widget]*overrideScope)
+
+type overrideScope struct {
+	th      fyne.Theme
+	cacheID string
+}
+
+var nextID = 1
+
+// OverrideTheme allows an app to specify that a single object should use a different theme to the app.
+// This should be used sparingly to avoid a jarring user experience.
+// If the object is a container it will theme the children, if it is a canvas primitive it will do nothing.
+//
+// Since: 2.5
+func OverrideTheme(o fyne.CanvasObject, th fyne.Theme) {
+	s := &overrideScope{th: th, cacheID: strconv.Itoa(nextID)}
+	overrideTheme(o, s, nextID)
+	nextID++
+}
+
+func WidgetTheme(o fyne.Widget) fyne.Theme {
+	data, ok := overrides[o]
+	if !ok {
+		return nil
+	}
+
+	return data.th
+}
+
+func OverrideResourceTheme(res fyne.Resource, w fyne.Widget) fyne.Resource {
+	if th, ok := res.(fyne.ThemedResource); ok {
+		return &WidgetResource{ThemedResource: th, Owner: w}
+	}
+
+	return res
+}
+
+func themeForResource(res fyne.Resource) fyne.Theme {
+	if th, ok := res.(*WidgetResource); ok {
+		if over, ok := overrides[th.Owner]; ok {
+			return over.th
+		}
+	}
+
+	return fyne.CurrentApp().Settings().Theme()
+}
+
+type WidgetResource struct {
+	fyne.ThemedResource
+	Owner fyne.Widget
+}
+
+// Content returns the underlying content of the resource adapted to the current text color.
+func (res *WidgetResource) Content() []byte {
+	name := res.ThemeColorName()
+	if name == "" {
+		name = "foreground"
+	}
+
+	th := themeForResource(res)
+	return svg.Colorize(res.ThemedResource.Content(), th.Color(name, fyne.CurrentApp().Settings().ThemeVariant()))
```

Comment:
OK you can ignore this particular comment because I thought of a way to resolve this in the general case here: https://github.com/fyne-io/fyne/pull/4558